File: tests/unit/optimizer/test_backwardpass_al.cpp
#include <catch2/catch_test_macros.hpp>

#include <algorithm>
#include <cmath>
#include <limits>
#include <random>
#include <vector>

#include <Eigen/Dense>

#include <saltro/optimizer/backwardpass.h>
#include <saltro/pybind/satellite.h>

using namespace saltro;

namespace {

Eigen::MatrixXd makeAttitudeTraj(const Eigen::Vector4d& att, int n_cols) {
	Eigen::MatrixXd traj(4, n_cols);
	for (int k = 0; k < n_cols; ++k) {
		traj.col(k) = att;
	}
	return traj;
}

struct TrajectoryData {
	Eigen::MatrixXd X;
	Eigen::MatrixXd U;
	Eigen::MatrixXd R;
	Eigen::MatrixXd V;
	Eigen::MatrixXd B;
	Eigen::MatrixXd S;
	Eigen::MatrixXd rho;
	Eigen::MatrixXd boresight;
	Eigen::MatrixXd attitude_target;
};

class BackwardPassALFixture {
public:
	explicit BackwardPassALFixture(int n = 6)
		: N(n),
		  settings(),
		  satellite(makeInertia(), settings),
		  reg(1e-8) {
		configureSatellite();
		configureSettings();
	}

	TrajectoryData makeTrajectory(bool force_active) {
		const int nx = satellite.stateDim();
		const int nu = satellite.controlDim();

		TrajectoryData td;
		td.X = Eigen::MatrixXd::Zero(nx, N);
		td.U = Eigen::MatrixXd::Zero(nu, std::max(0, N - 1));
		td.R = Eigen::MatrixXd::Zero(3, N);
		td.V = Eigen::MatrixXd::Zero(3, N);
		td.B = Eigen::MatrixXd::Zero(3, N);
		td.S = Eigen::MatrixXd::Zero(3, N);
		td.rho = Eigen::MatrixXd::Zero(1, N);
		td.boresight = Eigen::MatrixXd::Zero(3, N);

		Satellite::VecX x0 = Satellite::VecX::Zero(nx);
		x0.segment<3>(Satellite::AV_INDEX) = Eigen::Vector3d(0.01, -0.005, 0.008);
		x0.segment<4>(Satellite::QUAT_INDEX) = Eigen::Vector4d(1.0, 0.0, 0.0, 0.0);

		for (int k = 0; k < N; ++k) {
			td.X.col(k) = x0;
			if (force_active && k < N - 1) {
				td.U.col(k) << 0.15, -0.12, 0.10, 0.01, -0.01, 0.01;
			}

			td.R.col(k) = Eigen::Vector3d(7000e3, 0.0, 0.0);
			td.V.col(k) = Eigen::Vector3d(0.0, 7500.0, 0.0);
			td.B.col(k) = Eigen::Vector3d(2.5e-5, -1.5e-5, 3.0e-5);
			td.S.col(k) = Eigen::Vector3d(1.0, 0.1, -0.05).normalized();
			td.boresight.col(k) = Eigen::Vector3d::UnitX();
		}

		Eigen::Vector4d att;
		att << std::numeric_limits<double>::quiet_NaN(), 0.0, 0.0, 0.0;
		td.attitude_target = makeAttitudeTraj(att, N);

		if (force_active) {
			settings.constraints.wmax = 1e-4;
			settings.constraints.control_limit_scale = 1.0;
			settings.constraints.sun_limit_angle = 3.14159265358979323846;
			settings.constraints.u_max = Eigen::VectorXd::Constant(satellite.controlDim(), 1e-4);
		}

		return td;
	}

	std::vector<Eigen::VectorXd> collectConstraints(const TrajectoryData& td) const {
		std::vector<Eigen::VectorXd> c_list;
		c_list.reserve(static_cast<size_t>(N));

		for (int k = 0; k < N; ++k) {
			Eigen::VectorXd uk = Eigen::VectorXd::Zero(satellite.controlDim());
			if (k < td.U.cols()) {
				uk = td.U.col(k);
			}
			c_list.push_back(satellite.constraints(k, N, td.X.col(k), uk, td.S.col(k), settings.constraints));
		}
		return c_list;
	}

	static void makeZeroAug(
		const std::vector<Eigen::VectorXd>& c_list,
		std::vector<Eigen::VectorXd>& lambda_aug,
		std::vector<Eigen::VectorXd>& mu_aug
	) {
		lambda_aug.clear();
		mu_aug.clear();
		lambda_aug.reserve(c_list.size());
		mu_aug.reserve(c_list.size());
		for (const Eigen::VectorXd& c : c_list) {
			lambda_aug.push_back(Eigen::VectorXd::Zero(c.size()));
			mu_aug.push_back(Eigen::VectorXd::Zero(c.size()));
		}
	}

	bool runBackward(
		const TrajectoryData& td,
		const std::vector<Eigen::VectorXd>& lambda_aug,
		const std::vector<Eigen::VectorXd>& mu_aug,
		std::vector<Eigen::MatrixXd>& K,
		std::vector<Eigen::VectorXd>& d,
		Eigen::Vector2d& deltaV
	) const {
		initializeGains(K, d);
		deltaV.setZero();
		return optimizer::backwardPass(
			satellite,
			td.X,
			td.U,
			td.R,
			td.V,
			td.B,
			td.S,
			td.rho,
			td.boresight,
			td.attitude_target,
			settings,
			reg,
			K,
			d,
			deltaV,
			lambda_aug,
			mu_aug
		);
	}

	void initializeGains(std::vector<Eigen::MatrixXd>& K, std::vector<Eigen::VectorXd>& d) const {
		const int steps = std::max(0, N - 1);
		K.assign(static_cast<size_t>(steps), Eigen::MatrixXd::Zero(satellite.controlDim(), satellite.reducedStateDim()));
		d.assign(static_cast<size_t>(steps), Eigen::VectorXd::Zero(satellite.controlDim()));
	}

	int N;
	PlannerSettings settings;
	Satellite satellite;
	double reg;

private:
	static Eigen::Matrix3d makeInertia() {
		Eigen::Matrix3d J = Eigen::Matrix3d::Zero();
		J(0, 0) = 0.067;
		J(1, 1) = 0.071;
		J(2, 2) = 0.069;
		return J;
	}

	void configureSatellite() {
		satellite.addMTQ(Eigen::Vector3d::UnitX(), 0.2);
		satellite.addMTQ(Eigen::Vector3d::UnitY(), 0.2);
		satellite.addMTQ(Eigen::Vector3d::UnitZ(), 0.2);
		satellite.addRW(Eigen::Vector3d::UnitX(), 0.001, 1e-5, 0.0, 0.02);
		satellite.addRW(Eigen::Vector3d::UnitY(), 0.001, 1e-5, 0.0, 0.02);
		satellite.addRW(Eigen::Vector3d::UnitZ(), 0.001, 1e-5, 0.0, 0.02);
	}

	void configureSettings() {
		settings.disturbances.plan_for_aero = false;
		settings.disturbances.plan_for_gg = false;
		settings.disturbances.plan_for_srp = false;
		settings.disturbances.plan_for_prop = false;
		settings.disturbances.plan_for_gendist = false;
		settings.disturbances.plan_for_resdipole = false;
		settings.num_passes = 1;
		settings.passes[0].dt = 0.5;
		settings.passes[0].reg.reg_init = reg;
		settings.passes[0].reg.reg_scale = 10.0;
		settings.passes[0].reg.reg_max = 1e4;
	}
};

double gainDeltaNorm(const std::vector<Eigen::MatrixXd>& A, const std::vector<Eigen::MatrixXd>& B) {
	double acc = 0.0;
	const size_t n = std::min(A.size(), B.size());
	for (size_t i = 0; i < n; ++i) {
		acc += (A[i] - B[i]).squaredNorm();
	}
	return std::sqrt(acc);
}

double feedforwardDeltaNorm(const std::vector<Eigen::VectorXd>& A, const std::vector<Eigen::VectorXd>& B) {
	double acc = 0.0;
	const size_t n = std::min(A.size(), B.size());
	for (size_t i = 0; i < n; ++i) {
		acc += (A[i] - B[i]).squaredNorm();
	}
	return std::sqrt(acc);
}

} // namespace

TEST_CASE("backward_pass AL: runs with constraint-sized multipliers", "[backward_pass][al][sized]") {
	BackwardPassALFixture fixture(6);
	TrajectoryData td = fixture.makeTrajectory(false);
	const std::vector<Eigen::VectorXd> c_list = fixture.collectConstraints(td);

	std::vector<Eigen::VectorXd> lambda_aug;
	std::vector<Eigen::VectorXd> mu_aug;
	BackwardPassALFixture::makeZeroAug(c_list, lambda_aug, mu_aug);

	std::vector<Eigen::MatrixXd> K;
	std::vector<Eigen::VectorXd> d;
	Eigen::Vector2d deltaV = Eigen::Vector2d::Zero();

	const bool ok = fixture.runBackward(td, lambda_aug, mu_aug, K, d, deltaV);

	REQUIRE(ok);
	REQUIRE(K.size() == static_cast<size_t>(fixture.N - 1));
	REQUIRE(d.size() == static_cast<size_t>(fixture.N - 1));
	for (size_t k = 0; k < K.size(); ++k) {
		REQUIRE(K[k].allFinite());
		REQUIRE(d[k].allFinite());
	}
	REQUIRE(deltaV.allFinite());
}

TEST_CASE("backward_pass AL: active penalties modify gains/feedforward", "[backward_pass][al][active_effect]") {
	BackwardPassALFixture fixture(6);
	TrajectoryData td = fixture.makeTrajectory(true);
	const std::vector<Eigen::VectorXd> c_list = fixture.collectConstraints(td);

	bool any_active = false;
	for (int k = 0; k < fixture.N - 1; ++k) {
		if ((c_list[static_cast<size_t>(k)].array() > 0.0).any()) {
			any_active = true;
			break;
		}
	}
	REQUIRE(any_active);

	std::vector<Eigen::VectorXd> lambda_zero;
	std::vector<Eigen::VectorXd> mu_zero;
	BackwardPassALFixture::makeZeroAug(c_list, lambda_zero, mu_zero);

	std::vector<Eigen::VectorXd> lambda_aug;
	std::vector<Eigen::VectorXd> mu_aug;
	lambda_aug.reserve(c_list.size());
	mu_aug.reserve(c_list.size());
	for (const Eigen::VectorXd& c : c_list) {
		Eigen::VectorXd lam = Eigen::VectorXd::Zero(c.size());
		Eigen::VectorXd mu = Eigen::VectorXd::Zero(c.size());
		for (int i = 0; i < c.size(); ++i) {
			if (c(i) > 0.0) {
				lam(i) = 5.0;
				mu(i) = 25.0;
			}
		}
		lambda_aug.push_back(lam);
		mu_aug.push_back(mu);
	}

	std::vector<Eigen::MatrixXd> K0;
	std::vector<Eigen::VectorXd> d0;
	Eigen::Vector2d deltaV0 = Eigen::Vector2d::Zero();
	const bool ok0 = fixture.runBackward(td, lambda_zero, mu_zero, K0, d0, deltaV0);

	std::vector<Eigen::MatrixXd> K1;
	std::vector<Eigen::VectorXd> d1;
	Eigen::Vector2d deltaV1 = Eigen::Vector2d::Zero();
	const bool ok1 = fixture.runBackward(td, lambda_aug, mu_aug, K1, d1, deltaV1);

	REQUIRE(ok0);
	REQUIRE(ok1);
	const double k_diff = gainDeltaNorm(K1, K0);
	const double d_diff = feedforwardDeltaNorm(d1, d0);
	const double dv_diff = (deltaV1 - deltaV0).norm();
	REQUIRE((k_diff > 1e-12 || d_diff > 1e-12));
	REQUIRE(dv_diff > 1e-12);
}

TEST_CASE("backward_pass AL: size mismatch lists are ignored", "[backward_pass][al][size_mismatch]") {
	BackwardPassALFixture fixture(6);
	TrajectoryData td = fixture.makeTrajectory(true);
	const std::vector<Eigen::VectorXd> c_list = fixture.collectConstraints(td);

	std::vector<Eigen::VectorXd> lambda_zero;
	std::vector<Eigen::VectorXd> mu_zero;
	BackwardPassALFixture::makeZeroAug(c_list, lambda_zero, mu_zero);

	std::vector<Eigen::VectorXd> lambda_bad;
	std::vector<Eigen::VectorXd> mu_bad;
	lambda_bad.reserve(c_list.size());
	mu_bad.reserve(c_list.size());
	for (const Eigen::VectorXd& c : c_list) {
		lambda_bad.push_back(Eigen::VectorXd::Zero(c.size() + 1));
		mu_bad.push_back(Eigen::VectorXd::Ones(c.size() + 2));
	}

	std::vector<Eigen::MatrixXd> K0;
	std::vector<Eigen::VectorXd> d0;
	Eigen::Vector2d deltaV0 = Eigen::Vector2d::Zero();
	const bool ok0 = fixture.runBackward(td, lambda_zero, mu_zero, K0, d0, deltaV0);

	std::vector<Eigen::MatrixXd> Kb;
	std::vector<Eigen::VectorXd> db;
	Eigen::Vector2d deltaVb = Eigen::Vector2d::Zero();
	const bool okb = fixture.runBackward(td, lambda_bad, mu_bad, Kb, db, deltaVb);

	REQUIRE(ok0);
	REQUIRE(okb);
	REQUIRE(gainDeltaNorm(Kb, K0) < 1e-12);
	REQUIRE(feedforwardDeltaNorm(db, d0) < 1e-12);
	REQUIRE((deltaVb - deltaV0).norm() < 1e-12);
}

TEST_CASE("backward_pass AL: partial-horizon multiplier lists are accepted", "[backward_pass][al][partial_horizon]") {
	BackwardPassALFixture fixture(6);
	TrajectoryData td = fixture.makeTrajectory(true);
	const std::vector<Eigen::VectorXd> c_list = fixture.collectConstraints(td);

	std::vector<Eigen::VectorXd> lambda_short;
	std::vector<Eigen::VectorXd> mu_short;
	lambda_short.reserve(1);
	mu_short.reserve(1);

	Eigen::VectorXd lam = Eigen::VectorXd::Zero(c_list[0].size());
	Eigen::VectorXd mu = Eigen::VectorXd::Zero(c_list[0].size());
	for (int i = 0; i < c_list[0].size(); ++i) {
		if (c_list[0](i) > 0.0) {
			lam(i) = 2.0;
			mu(i) = 10.0;
		}
	}
	lambda_short.push_back(lam);
	mu_short.push_back(mu);

	std::vector<Eigen::MatrixXd> K;
	std::vector<Eigen::VectorXd> d;
	Eigen::Vector2d deltaV = Eigen::Vector2d::Zero();

	const bool ok = fixture.runBackward(td, lambda_short, mu_short, K, d, deltaV);

	REQUIRE(ok);
	for (size_t k = 0; k < K.size(); ++k) {
		REQUIRE(K[k].allFinite());
		REQUIRE(d[k].allFinite());
	}
	REQUIRE(deltaV.allFinite());
}

TEST_CASE("backward_pass AL: longer horizon random nonnegative multipliers remain stable", "[backward_pass][al][stability]") {
	BackwardPassALFixture fixture(10);
	TrajectoryData td = fixture.makeTrajectory(true);
	const std::vector<Eigen::VectorXd> c_list = fixture.collectConstraints(td);

	std::mt19937 rng(7);
	std::normal_distribution<double> dist_lam(0.2, 0.1);
	std::normal_distribution<double> dist_mu(5.0, 2.0);

	std::vector<Eigen::VectorXd> lambda_aug;
	std::vector<Eigen::VectorXd> mu_aug;
	lambda_aug.reserve(c_list.size());
	mu_aug.reserve(c_list.size());

	for (const Eigen::VectorXd& c : c_list) {
		Eigen::VectorXd lam(c.size());
		Eigen::VectorXd mu(c.size());
		for (int i = 0; i < c.size(); ++i) {
			lam(i) = std::max(0.0, dist_lam(rng));
			mu(i) = std::max(0.0, dist_mu(rng));
		}
		lambda_aug.push_back(lam);
		mu_aug.push_back(mu);
	}

	std::vector<Eigen::MatrixXd> K;
	std::vector<Eigen::VectorXd> d;
	Eigen::Vector2d deltaV = Eigen::Vector2d::Zero();

	const bool ok = fixture.runBackward(td, lambda_aug, mu_aug, K, d, deltaV);

	REQUIRE(ok);
	REQUIRE(K.size() == static_cast<size_t>(fixture.N - 1));
	REQUIRE(d.size() == static_cast<size_t>(fixture.N - 1));
	for (size_t k = 0; k < K.size(); ++k) {
		REQUIRE(K[k].allFinite());
		REQUIRE(d[k].allFinite());
	}
	REQUIRE(deltaV.allFinite());
}

File: tests/unit/optimizer/test_backwardpass_al.py
"""
Python tests for backward pass with augmented Lagrangian inputs.
"""

import numpy as np
import pytest
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[3]
sys.path.insert(0, str(ROOT / "build"))
import saltro_py


def make_attitude_traj(att, n_cols):
    traj = np.zeros((4, n_cols))
    for k in range(n_cols):
        traj[:, k] = att
    return traj


class BackwardPassALFixture:
    def __init__(self, n=6):
        self.N = n
        self.settings = saltro_py.PlannerSettings()

        J = np.diag([0.067, 0.071, 0.069])
        self.satellite = saltro_py.Satellite(J, self.settings)

        self.satellite.addMTQ(np.array([1.0, 0.0, 0.0]), 0.2)
        self.satellite.addMTQ(np.array([0.0, 1.0, 0.0]), 0.2)
        self.satellite.addMTQ(np.array([0.0, 0.0, 1.0]), 0.2)
        self.satellite.addRW(np.array([1.0, 0.0, 0.0]), 0.001, 1e-5, 0.0, 0.02)
        self.satellite.addRW(np.array([0.0, 1.0, 0.0]), 0.001, 1e-5, 0.0, 0.02)
        self.satellite.addRW(np.array([0.0, 0.0, 1.0]), 0.001, 1e-5, 0.0, 0.02)

        self.settings.disturbances.plan_for_aero = False
        self.settings.disturbances.plan_for_gg = False
        self.settings.disturbances.plan_for_srp = False
        self.settings.disturbances.plan_for_prop = False
        self.settings.disturbances.plan_for_gendist = False
        self.settings.disturbances.plan_for_resdipole = False
        self.settings.num_passes = 1
        self.settings.passes[0].dt = 0.5
        self.settings.passes[0].reg.reg_init = 1e-8
        self.settings.passes[0].reg.reg_scale = 10.0
        self.settings.passes[0].reg.reg_max = 1e4

        self.reg = self.settings.passes[0].reg.reg_init

    def make_trajectory(self, force_active=False):
        nx = self.satellite.stateDim
        nu = self.satellite.controlDim
        N = self.N

        X = np.zeros((nx, N))
        U = np.zeros((nu, N - 1))

        x0 = np.zeros(nx)
        x0[0:3] = np.array([0.01, -0.005, 0.008])
        x0[3:7] = np.array([1.0, 0.0, 0.0, 0.0])

        for k in range(N):
            X[:, k] = x0
            if force_active and k < N - 1:
                U[:, k] = np.array([0.15, -0.12, 0.10, 0.01, -0.01, 0.01])

        if force_active:
            self.settings.constraints.wmax = 1e-4
            self.settings.constraints.control_limit_scale = 1.0
            self.settings.constraints.sun_limit_angle = np.pi

        R = np.zeros((3, N))
        V = np.zeros((3, N))
        B = np.zeros((3, N))
        S = np.zeros((3, N))
        rho = np.zeros((1, N))
        boresight = np.zeros((3, N))

        for k in range(N):
            R[:, k] = np.array([7000e3, 0.0, 0.0])
            V[:, k] = np.array([0.0, 7500.0, 0.0])
            B[:, k] = np.array([2.5e-5, -1.5e-5, 3.0e-5])
            S[:, k] = np.array([1.0, 0.1, -0.05])
            S[:, k] /= np.linalg.norm(S[:, k])
            boresight[:, k] = np.array([1.0, 0.0, 0.0])

        attitude_target = np.array([np.nan, 0.0, 0.0, 0.0])
        attitude_target_traj = make_attitude_traj(attitude_target, N)

        return X, U, R, V, B, S, rho, boresight, attitude_target_traj


def collect_constraints(fixture, X, U, S):
    sat = fixture.satellite
    cfg = fixture.settings.constraints
    N = X.shape[1]
    nu = sat.controlDim
    c_list = []
    for k in range(N):
        uk = U[:, k] if k < U.shape[1] else np.zeros(nu)
        ck = np.asarray(sat.constraints(k, N, X[:, k], uk, S[:, k], cfg), dtype=float)
        c_list.append(ck)
    return c_list


def make_zero_aug(c_list):
    lambda_aug = [np.zeros_like(c) for c in c_list]
    mu_aug = [np.zeros_like(c) for c in c_list]
    return lambda_aug, mu_aug


def run_backward(fixture, X, U, R, V, B, S, rho, boresight, attitude_target, lambda_aug, mu_aug):
    return saltro_py.backward_pass(
        fixture.satellite,
        X,
        U,
        R,
        V,
        B,
        S,
        rho,
        boresight,
        attitude_target,
        fixture.settings,
        lambda_aug,
        mu_aug,
        fixture.reg,
    )


@pytest.fixture
def fixture():
    return BackwardPassALFixture(n=6)


def test_backward_pass_al_runs_with_constraint_sized_multipliers(fixture):
    X, U, R, V, B, S, rho, boresight, attitude_target = fixture.make_trajectory(force_active=False)
    c_list = collect_constraints(fixture, X, U, S)
    lambda_aug, mu_aug = make_zero_aug(c_list)

    ok, K, d, deltaV = run_backward(fixture, X, U, R, V, B, S, rho, boresight, attitude_target, lambda_aug, mu_aug)

    assert ok
    assert K.shape == (fixture.N - 1, fixture.satellite.controlDim, fixture.satellite.reducedStateDim)
    assert d.shape == (fixture.satellite.controlDim, fixture.N - 1)
    assert np.all(np.isfinite(K))
    assert np.all(np.isfinite(d))
    assert np.all(np.isfinite(deltaV))


def test_backward_pass_al_active_penalties_modify_solution(fixture):
    X, U, R, V, B, S, rho, boresight, attitude_target = fixture.make_trajectory(force_active=True)
    c_list = collect_constraints(fixture, X, U, S)
    zero_lambda, zero_mu = make_zero_aug(c_list)

    any_active = any(np.any(c > 0.0) for c in c_list[:-1])
    assert any_active

    lambda_aug = [np.where(c > 0.0, 5.0, 0.0) for c in c_list]
    mu_aug = [np.where(c > 0.0, 25.0, 0.0) for c in c_list]

    ok0, K0, d0, deltaV0 = run_backward(fixture, X, U, R, V, B, S, rho, boresight, attitude_target, zero_lambda, zero_mu)
    ok1, K1, d1, deltaV1 = run_backward(fixture, X, U, R, V, B, S, rho, boresight, attitude_target, lambda_aug, mu_aug)

    assert ok0 and ok1
    assert np.linalg.norm(K1 - K0) > 1e-12 or np.linalg.norm(d1 - d0) > 1e-12
    assert np.linalg.norm(deltaV1 - deltaV0) > 1e-12


def test_backward_pass_al_size_mismatch_is_ignored(fixture):
    X, U, R, V, B, S, rho, boresight, attitude_target = fixture.make_trajectory(force_active=True)
    c_list = collect_constraints(fixture, X, U, S)
    zero_lambda, zero_mu = make_zero_aug(c_list)

    bad_lambda = [np.zeros(c.shape[0] + 1) for c in c_list]
    bad_mu = [np.ones(c.shape[0] + 2) for c in c_list]

    ok0, K0, d0, deltaV0 = run_backward(fixture, X, U, R, V, B, S, rho, boresight, attitude_target, zero_lambda, zero_mu)
    okb, Kb, db, deltaVb = run_backward(fixture, X, U, R, V, B, S, rho, boresight, attitude_target, bad_lambda, bad_mu)

    assert ok0 and okb
    assert np.allclose(Kb, K0, atol=1e-12, rtol=1e-10)
    assert np.allclose(db, d0, atol=1e-12, rtol=1e-10)
    assert np.allclose(deltaVb, deltaV0, atol=1e-12, rtol=1e-10)


def test_backward_pass_al_partial_horizon_multiplier_lists(fixture):
    X, U, R, V, B, S, rho, boresight, attitude_target = fixture.make_trajectory(force_active=True)
    c_list = collect_constraints(fixture, X, U, S)

    lambda_short = [np.where(c_list[0] > 0.0, 2.0, 0.0)]
    mu_short = [np.where(c_list[0] > 0.0, 10.0, 0.0)]

    ok, K, d, deltaV = run_backward(
        fixture, X, U, R, V, B, S, rho, boresight, attitude_target, lambda_short, mu_short
    )

    assert ok
    assert np.all(np.isfinite(K))
    assert np.all(np.isfinite(d))
    assert np.all(np.isfinite(deltaV))


def test_backward_pass_al_long_horizon_random_multipliers_are_stable():
    fixture = BackwardPassALFixture(n=10)
    X, U, R, V, B, S, rho, boresight, attitude_target = fixture.make_trajectory(force_active=True)

    rng = np.random.default_rng(7)
    c_list = collect_constraints(fixture, X, U, S)

    lambda_aug = []
    mu_aug = []
    for c in c_list:
        lambda_aug.append(np.maximum(0.0, rng.normal(0.2, 0.1, size=c.shape[0])))
        mu_aug.append(np.maximum(0.0, rng.normal(5.0, 2.0, size=c.shape[0])))

    ok, K, d, deltaV = run_backward(
        fixture, X, U, R, V, B, S, rho, boresight, attitude_target, lambda_aug, mu_aug
    )

    assert ok
    assert K.shape[0] == fixture.N - 1
    assert d.shape[1] == fixture.N - 1
    assert np.all(np.isfinite(K))
    assert np.all(np.isfinite(d))
    assert np.all(np.isfinite(deltaV))
